src/Avalonia.Controls/Primitives/ScrollBar.cs

```diff
@@ -178,7 +185,58 @@ private void UpdateIsVisible()
                 _ => throw new InvalidOperationException("Invalid value for ScrollBar.Visibility.")
             };
 
-            SetValue(IsVisibleProperty, isVisible);
+            SetCurrentValue(IsVisibleProperty, isVisible);
+        }
+
+        protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
+        {
+            base.OnAttachedToVisualTree(e);
+            AttachToScrollViewer();
+        }
+
+        /// <summary>
+        /// Locates the first <see cref="ScrollViewer"/> ancestor and binds to its properties. Properties which have been set through other means are not bound.
+        /// </summary>
+        /// <remarks>
+        /// This method is automatically called when the control is attached to a visual tree.
+        /// </remarks>
+        protected internal virtual void AttachToScrollViewer()
```

Comment:
The nice part of doing things this way round is that `ScrollViewer` has only the properties that you would expect, and it is `ScrollBar` which reads them and does the "implementation details" bit of transforming them to suit its own needs. This provides better separation than before.

I don't think that a separate "properties object" is the way to go, because none of the properties in question are implementation details of the template. They are all semantic properties that I would expect to find on the `ScrollViewer` itself.

One thing that could be done within the scope of this PR is moving the code that talks to `ScrollViewer` into a more specific class, which would have a name like "ScrollViewerScrollBar" (slightly silly I know). In the long run I also like the idea of a shared controller object.